Auto add techdocs core plugin

Xantier opened the pull request:
## Hey, I just made a Pull Request!

Add techdocs-core plugin to techdocs builder automatically

Currently users need to modify their mkdocs files to container techdocs-core plugin manually when they want to expose techdocs in Backstage. To have a standardized set of extensions without the need to modify existing files (and possibly pollute alternative/additional mkdocs pipelines) this should be added automatically.

This PR adds a helper function to modify the mkdocs file to contain this plugin if it doesn't exist in the config file.
#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [n/a] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

Rugvip commented:
@backstage/techdocs-core, will let you take the first look at this one

andrewthauer reviewed (commented):
I noticed there is some efforts going on to de-couple from mkdocs. So it's possible the new config option to omit would be obsolete for non mkdocs use cases. However, I'm glad this was considered, because we are currently not using the core plugin. Seems like the config option could just be ignored for non mkdocs or maybe it would make sense introduce a sub type for mkdocs specifically similar to different storage provider's?

iamEAP commented on packages/techdocs-common/src/stages/generate/types.ts:
```diff
@@ -29,6 +29,7 @@ export type GeneratorConfig = {
   runIn: GeneratorRunInType;
   dockerImage?: string;
   pullImage?: boolean;
+  omitTechdocsPlugin?: boolean;
```
Wonder if this could be a little less ambiguously named?  Like `omitTechdocsCoreMkdocsPlugin` or something?  Or, to build off of @andrewthauer's comment, nesting it, perhaps: `techdocs.generator.mkdocs.omitTechdocsCorePlugin` or something?

Xantier replied:
Yup, that makes sense. Modifying.

iamEAP commented on packages/techdocs-common/src/stages/generate/helpers.ts:
```diff
@@ -298,6 +266,87 @@ export const patchMkdocsYmlPreBuild = async (
   }
 };
 
+/**
```
Wonder if it makes sense to separate out these yml (and perhaps even `techdocs_metadata.json` transformations out of `helpers` and into like a...  Transformer file or folder?  😐

Xantier replied:
I moved these now to initially reside in mkdocs specific 'patchers' file, since that's what they are called.

iamEAP reviewed (commented):
Thanks @Xantier.  This aligns well with reducing onboarding friction.

Couple of overall thoughts on this:

- The config-based approach for omitting this transformation works well for docs generated on-the-fly by Backstage, but how would someone generating docs via `@techdocs/cli` avoid this transformation?
- As part of #9636, we're considering making this type of build-time transformation extensible (more in the `techdocs_metadata.json` realm, but `mkdocs.yml` transformation is easy to imagine too).  ...More just a heads up that we'd love for you to read through and provide thoughts/feedback.  But perhaps it might influence this in some way.
- There might be a couple more spots in documentation that should be updated?  I'm thinking in particular of the [Creating and Publishing Documentation](https://backstage.io/docs/features/techdocs/creating-and-publishing) section, but there could be others.

Xantier commented:
Thanks @iamEAP,

I've addressed the comments on this for now. AFAICS the mentioned plugin system would likely keep this unaffected unless we want to externalize this eventually and add it in as a pluggable item. Might be too soon to think about that though at this point.

emmaindal commented on docs/features/techdocs/creating-and-publishing.md:
````diff
@@ -79,6 +79,8 @@ plugins:
   - techdocs-core
 ```
 
+The plugins section above is optional. Backstage automatically adds the `techdocs-core` plugin to the mkdocs file if it is missing. This functionality can be turned off with a configuration option in Backstage.
````
Maybe make this a "note" to give it a bit more attention? And also point to the docs of how its turned off. Something like:

```markdown
 > Note - The plugins section above is optional. Backstage automatically adds the `techdocs-core` plugin to the 
 > mkdocs file if it is missing. This functionality can be turned off with a [configuration option](./configuration.md) in Backstage.
```

Xantier replied:
That makes sense yeah. Updated.

emmaindal reviewed (approved):
This looks good to me! Just one smaller comment around the documentation that I think we could get in before merging!

emmaindal commented:
@backstage/maintainers need your review too!

Rugvip reviewed (approved):
🟢  to unblock changeset and docs
